### DISCUSSION TITLE
Minimum SDK Error While Building Android

### DISCUSSION THREAD

**Discussion by VikasAgrawal-04** [codemagic-ci-cd/codemagic-docs#2673] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Create Build Normally Using codemagic. 
![Screenshot from 2024-08-07 19-48-38](https://github.com/user-attachments/assets/1e330529-e586-4287-9e07-59ab712dd9dc)

### Expected results

It should have created the build and submitted it to Play Store

### Actual results

It gives me error  -:  Failed to sign artifact file_cp_resources.apk. Failed to sign apk: Exception in thread "main" com.android.apksig.apk.MinSdkVersionException: Failed to determine APK's minimum supported platform version. Use --min-sdk-version to override

**Comment by himesh-cm** [comment]
Hey @VikasAgrawal-04 
Can you please share the build ID with me? You can find it under your build details page's 'ID' section.  ﻿Navigate to the 'Builds' section on the left-hand menu, select the build you want to view, and click the copy icon.

**Reply by VikasAgrawal-04** [reply]
This is the ID 66b33e44856723bafc1e7fe0

**Reply by VikasAgrawal-04** [reply]
Please update how to rectify this????

**Reply by himesh-cm** [reply]
Hey @VikasAgrawal-04 
Can you check your app-level `build.gradle` file and make sure that the minimum SDK version is specified?

```
android {
    defaultConfig {
        minSdkVersion 21  // Or whatever your minimum SDK version is
        targetSdkVersion 30
    }
}
```

**Reply by VikasAgrawal-04** [reply]
I already have declared that in build.gradle


![PXL_20240813_102853894.jpg](https://github.com/user-attachments/assets/e9f27408-ee68-4221-b00a-245eed17d876)

**Reply by sikha24** [reply]
Facing the same issue please suggest

**Reply by himesh-cm** [reply]
Can you please confirm if you have followed the Gradle setup as mentioned here in the docs?
https://docs.codemagic.io/yaml-code-signing/signing-android/#signing-android-apps-using-gradle

**Reply by VikasAgrawal-04** [reply]
This is the Build ID 66b33e44856723bafc1e7fe0

**Reply by VikasAgrawal-04** [reply]
> Can you please confirm if you have followed the Gradle setup as mentioned here in the docs?
> https://docs.codemagic.io/yaml-code-signing/signing-android/#signing-android-apps-using-gradle

This is the ID 66b33e44856723bafc1e7fe0

**Reply by himesh-cm** [reply]
Please share your `android/app/build.gradle` for further debugging.
If there is any sensitive information, please make sure to remove it.

**Reply by VikasAgrawal-04** [reply]
> Please share your `android/app/build.gradle` for further debugging. If there is any sensitive information, please make sure to remove it.

![image](https://github.com/user-attachments/assets/fad5bf26-7479-4cad-b9ff-217b6a37240b)
![image](https://github.com/user-attachments/assets/c7ec46dc-6903-4105-9b30-8fd06808d7ad)
![image](https://github.com/user-attachments/assets/4088a0dd-6674-466c-baa7-372641eb8b5e)

**Reply by VikasAgrawal-04** [reply]
> Please share your `android/app/build.gradle` for further debugging. If there is any sensitive information, please make sure to remove it.

def localProperties = new Properties()
def localPropertiesFile = rootProject.file('local.properties')
if (localPropertiesFile.exists()) {
    localPropertiesFile.withReader('UTF-8') { reader ->
        localProperties.load(reader)
    }
}

def flutterRoot = localProperties.getProperty('flutter.sdk')
if (flutterRoot == null) {
    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
}

def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
if (flutterVersionCode == null) {
    flutterVersionCode = '1'
}

def flutterVersionName = localProperties.getProperty('flutter.versionName')
if (flutterVersionName == null) {
    flutterVersionName = '1.0'
}

apply plugin: 'com.android.application'
apply plugin: 'kotlin-android'
apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
apply plugin: 'com.google.gms.google-services'
apply plugin: 'com.google.firebase.crashlytics'

def keystoreProperties = new Properties()
def keystorePropertiesFile = rootProject.file('key.properties')
if (keystorePropertiesFile.exists()) {
    keystoreProperties.load(new FileInputStream(keystorePropertiesFile))
}


android {
    ndkVersion flutter.ndkVersion
    compileSdkVersion 34

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    kotlinOptions {
        jvmTarget = '1.8'
    }

    sourceSets {
        main.java.srcDirs += 'src/main/kotlin'
    }

    defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId "com.opmaint.opmaint"
        // You can update the following values to match your application needs.
        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-build-configuration.
        minSdkVersion 21
        targetSdkVersion 34
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
        multiDexEnabled true
    }

    signingConfigs {
            release {
               keyAlias keystoreProperties['keyAlias']
               keyPassword keystoreProperties['keyPassword']
               storeFile keystoreProperties['storeFile'] ? file(keystoreProperties['storeFile']) : null
               storePassword keystoreProperties['storePassword']
            }
    }


    buildTypes {
        release {
            // TODO: Add your own signing config for the release build.
            // Signing with the debug keys for now, so `flutter run --release` works.
            signingConfig signingConfigs.debug
            signingConfig signingConfigs.release
        }
    }

    packagingOptions {
        pickFirst 'lib/arm64-v8a/libc++_shared.so'
    }
}

flutter {
    source '../..'
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation platform('com.google.firebase:firebase-bom:31.2.0')
    implementation 'com.google.firebase:firebase-analytics-ktx'
}

**Reply by VikasAgrawal-04** [reply]
> Please share your `android/app/build.gradle` for further debugging. If there is any sensitive information, please make sure to remove it.

Hey could you look into that issue. It's been a while. I guess I have provided you everything.

**Reply by himesh-cm** [reply]
When you run `flutter build appbundle --release` locally after cleaning caches, do you encounter a similar error? Are you using the same version of Flutter locally and on codemagic?

**Reply by VikasAgrawal-04** [reply]
> When you run `flutter build appbundle --release` locally after cleaning caches, do you encounter a similar error? Are you using the same version of Flutter locally and on codemagic?

It works fine when I run it locally. No error encountered there.

**Reply by himesh-cm** [reply]
Are you using the same version of Flutter locally and on codemagic?

**Reply by VikasAgrawal-04** [reply]
Yes exact same.

**Reply by himesh-cm** [reply]
Hey @VikasAgrawal-04, Please share your email ID. Let's have a one-on-one screen-sharing call to discuss this further.

**Reply by himesh-cm** [reply]
email sent

**Reply by himesh-cm** [reply]
I see that you have booked a slot for `05:00pm - Wednesday, 4 September 2024 (India Standard Time)`, If this is urgent for you i suggest you book a slot for tomorrow.

**Comment by himesh-cm** [comment]
I am closing this thread for now and will update the thread once we have a resolution.

**Comment by VikasAgrawal-04** [comment]
Hey Himesh,
Do we have a solution? Will we be connecting today?

**Comment by kevin4dhd** [comment]
any solution?
